File: Practice/leetcode/problem_1_10/src/add_two_numbers.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>,
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode { next: None, val }
  }
}

pub fn add_two_numbers(
  l1: Option<Box<ListNode>>,
  l2: Option<Box<ListNode>>,
) -> Option<Box<ListNode>> {
  let mut l1_current = l1;
  let mut l2_current = l2;

  let mut head = Some(Box::new(ListNode::new(0)));
  let mut current = head.as_mut();
  let mut carry = 0;

  while l1_current.is_some() || l2_current.is_some() {
    let mut sum = carry;
    if let Some(node) = l1_current {
      sum += node.val;
      l1_current = node.next;
    }
    if let Some(node) = l2_current {
      sum += node.val;
      l2_current = node.next;
    }

    carry = sum / 10;

    if let Some(node) = current {
      node.next = Some(Box::new(ListNode::new(sum % 10)));
      current = node.next.as_mut();
    }
  }

  if carry > 0 {
    current.unwrap().next = Some(Box::new(ListNode::new(carry)));
  }

  return head.unwrap().next;
}

File: Practice/leetcode_Go/problem_1_10/add_two_numbers/add_two_numbers.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	result := &ListNode{0, nil}
	lResult := result
	var quotient = 0

	for l1 != nil || l2 != nil {
		var val1, val2 = 0, 0
		if l1 != nil {
			val1 = l1.Val
		}
		if l2 != nil {
			val2 = l2.Val
		}

		sum := val1 + val2 + quotient

		quotient = sum / 10
		lResult.Next = &ListNode{sum % 10, nil}
		lResult = lResult.Next

		if l1 != nil {
			l1 = l1.Next
		}

		if l2 != nil {
			l2 = l2.Next
		}
	}

	if quotient > 0 {
		lResult.Next = &ListNode{quotient, nil}
	}
	return result.Next
}
func main() {
	l1 := ListNode{2, nil}
	l1.Next = &ListNode{4, nil}
	l1.Next.Next = &ListNode{6, nil}

	l2 := ListNode{5, nil}
	l2.Next = &ListNode{6, nil}
	l2.Next.Next = &ListNode{4, nil}

	addTwoNumbers(&l1, &l2)
}
